src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,584 @@
+            self.queue,
+            now.elapsed(),
+            self.stale_files.read().len(),
+        );
+        Ok(())
+    }
+
+    /// Generates a Fake log used for recycling.
+    ///
+    /// Attention, this function is only called when `Config.enable-log-recycle`
+    /// is true.
+    fn build_stale_file(
+        file_system: &F,
+        paths: &Paths,
+        file_id: FileId,
+        format: LogFileFormat,
+        target_file_size: usize,
+    ) -> Result<FileWithFormat<F>> {
+        let path_id = FileCollection::<F>::get_valid_path(paths, target_file_size);
+        let file_path = file_id.build_stale_file_path(&paths[path_id]);
+        let fd = Arc::new(file_system.create(&file_path)?);
+        let mut writer = file_system.new_writer(fd.clone())?;
+        let mut written = 0_usize;
+        let buf = vec![0; std::cmp::min(LOG_STALE_FILE_BUF_SIZE, target_file_size)];
+        while written <= target_file_size {
+            writer.write_all(&buf).unwrap_or_else(|e| {
+                panic!("failed to prepare stale file: {}", e);
+            });
+            written += buf.len();
+        }
+        // Metadata of stale files are not what we're truely concerned. So,
+        // they can be ignored by clear them here.
+        Ok(FileWithFormat {
+            handle: fd,
+            format,
+            path_id,
+        })
+    }
+
+    /// Returns a valid path for dumping new files.
+    ///
+    /// Default: 0 -> default index of the main directory.
+    pub fn get_valid_path(paths: &Paths, _target_file_size: usize) -> PathId {
+        debug_assert!(paths.len() <= 2);
+        0
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::VecDeque;
+    use tempfile::Builder;
+
+    use super::super::format::{FileNameExt, LogFileFormat};
+    use super::*;
+    use crate::env::DefaultFileSystem;
+    use crate::pipe_log::Version;
+    use crate::util::ReadableSize;
+
+    #[test]
+    fn test_file_collection() {
+        fn new_file_handler(
+            path: &str,
+            file_id: FileId,
+            version: Version,
+            is_stale: bool,
+        ) -> FileWithFormat<DefaultFileSystem> {
+            let file_path = if !is_stale {
+                file_id.build_file_path(path)
+            } else {
+                file_id.build_stale_file_path(path)
+            };
+            FileWithFormat {
+                handle: Arc::new(DefaultFileSystem.create(&file_path).unwrap()),
+                format: LogFileFormat::new(version, 0 /* alignment */),
+                path_id: 0, /* default with Main dir */
+            }
+        }
+        let dir = Builder::new()
+            .prefix("test_file_collection")
+            .tempdir()
+            .unwrap();
+        let path = String::from(dir.path().to_str().unwrap());
+        let target_file_size = ReadableSize(1);
+        // | 12
+        let mut active_files = FileList::new(
```

Comment:
Either test `FileList`, or test `FileCollection`, Avoid testing two concepts in one case.